File: server/app/services/workflow.py
from app.models.enums import UserRole, DrawingStatus

WORKFLOW_TRANSITIONS = {
    DrawingStatus.UNASSIGNED: {
        "ASSIGN": {
            "role": UserRole.ADMIN,
            "next": DrawingStatus.DRAFTING,
            "lock": True,
        }
    },
    DrawingStatus.DRAFTING: {
        "CLAIM": {
            "role": UserRole.DRAFTER,
            "next": DrawingStatus.DRAFTING,
            "lock": True,
        },
        "SUBMIT": {
            "role": UserRole.DRAFTER,
            "next": DrawingStatus.FIRST_QC,
            "lock": False,
        },
    },
    DrawingStatus.FIRST_QC: {
        "CLAIM": {
            "role": UserRole.SHIFT_LEAD,
            "next": DrawingStatus.FIRST_QC,
            "lock": True,
        },
        "SUBMIT": {
            "role": UserRole.SHIFT_LEAD,
            "next": DrawingStatus.FINAL_QC,
            "lock": False,
        },
    },
    DrawingStatus.FINAL_QC: {
        "CLAIM": {
            "role": UserRole.FINAL_QC,
            "next": DrawingStatus.FINAL_QC,
            "lock": True,
        },
        "APPROVE": {
            "role": UserRole.FINAL_QC,
            "next": DrawingStatus.APPROVED,
            "lock": False,
        },
    },
}

File: client/src/utils/workflow.ts
export type Role =
  | "ADMIN"
  | "DRAFTER"
  | "SHIFT_LEAD"
  | "FINAL_QC";

export type Action = "ASSIGN" | "CLAIM" | "SUBMIT" | "APPROVE";

export function getAvailableActions(
  role: Role,
  status: string,
  isAssignedToMe: boolean
): Action[] {
  // UNASSIGNED → only ADMIN can ASSIGN
  if (status === "UNASSIGNED" && role === "ADMIN") {
    return ["ASSIGN"];
  }

  // DRAFTING
  if (status === "DRAFTING" && role === "DRAFTER") {
    return isAssignedToMe ? ["SUBMIT"] : ["CLAIM"];
  }

  // FIRST QC
  if (status === "FIRST_QC" && role === "SHIFT_LEAD") {
    return isAssignedToMe ? ["SUBMIT"] : ["CLAIM"];
  }

  // FINAL QC
  if (status === "FINAL_QC" && role === "FINAL_QC") {
    return isAssignedToMe ? ["APPROVE"] : ["CLAIM"];
  }

  return [];
}
